File: sdk/csharp/src/Providers/ThrowawayMail.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// ThrowawayMail 渠道 — https://throwawaymail.app（API: /api）。
/// 创建：POST /mailboxes 返回 mailbox_id 与 address。
/// 收信：GET /mailboxes/{id}/messages 列表 + 逐封 GET /mailboxes/{id}/messages/{mid} 取详情。
/// </summary>
public static class ThrowawayMail
{
    private const string ApiBase = "https://throwawaymail.app/api";
    private static readonly Dictionary<string, string> Headers = new() { ["Accept"] = "application/json" };

    private static Dictionary<string, object?> MessageRaw(JsonObject raw, string recipient)
    {
        var to = recipient;
        if (raw["to_addresses"] is JsonArray ta && ta.Count > 0)
            to = Json.NodeToString(ta[0]);
        return new Dictionary<string, object?>
        {
            ["id"] = Json.Str(raw, "message_id"),
            ["messageId"] = Json.Str(raw, "message_id"),
            ["from_address"] = Json.Str(raw, "from_address"),
            ["fromName"] = Json.Str(raw, "from_name"),
            ["to"] = to,
            ["subject"] = Json.Str(raw, "subject"),
            ["received_at"] = Json.Str(raw, "received_at"),
            ["read"] = (raw["read"] as JsonValue)?.GetValue<bool>() ?? false,
            ["text"] = Json.Str(raw, "text"),
            ["html"] = Json.Str(raw, "html"),
        };
    }

    /// <summary>创建临时邮箱</summary>
    public static EmailInfo Generate()
    {
        var resp = Http.Post($"{ApiBase}/mailboxes", null, null, new Dictionary<string, string>(Headers));
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        var mailboxId = Json.Str(data, "mailbox_id").Trim();
        var email = Json.Str(data, "address").Trim();
        if (string.IsNullOrEmpty(mailboxId) || string.IsNullOrEmpty(email) || !email.Contains('@'))
            throw new Exception("throwawaymail: invalid mailbox response");
        long? expires = null;
        if (data!.TryGetPropertyValue("expires_at", out var ea) && ea is not null && long.TryParse(ea.ToString(), out var ev)) expires = ev;
        return new EmailInfo("throwawaymail", email, mailboxId, expires, data["created_at"]?.ToString());
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string? token, string email)
    {
        var mailboxId = (token ?? "").Trim();
        var address = (email ?? "").Trim();
        if (string.IsNullOrEmpty(mailboxId)) throw new Exception("throwawaymail: empty mailbox id");
        if (string.IsNullOrEmpty(address)) throw new Exception("throwawaymail: empty email");

        var resp = Http.Get($"{ApiBase}/mailboxes/{WebUtility.UrlEncode(mailboxId)}/messages", new Dictionary<string, string>(Headers));
        resp.EnsureSuccess();
        var rows = Json.Parse(resp.Body) as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;

        foreach (var r in rows)
        {
            if (r is not JsonObject item) continue;
            var messageId = Json.Str(item, "message_id").Trim();
            var detail = string.IsNullOrEmpty(messageId) ? null : FetchMessage(mailboxId, messageId);
            result.Add(Normalize.NormalizeEmail(MessageRaw(detail ?? item, address), address));
        }
        return result;
    }

    private static JsonObject? FetchMessage(string mailboxId, string messageId)
    {
        var resp = Http.Get($"{ApiBase}/mailboxes/{WebUtility.UrlEncode(mailboxId)}/messages/{WebUtility.UrlEncode(messageId)}",
            new Dictionary<string, string>(Headers));
        if (!resp.Ok) return null;
        return Json.Parse(resp.Body) as JsonObject;
    }
}

File: sdk/php/src/Providers/ThrowawayMail.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * throwawaymail.app 渠道实现
 *
 * API: https://throwawaymail.app/api
 * 流程: POST /mailboxes 创建邮箱（mailbox_id 作为 token）→
 *       GET /mailboxes/{id}/messages 列表 → GET /mailboxes/{id}/messages/{mid} 详情。
 */
final class ThrowawayMail
{
    private const CHANNEL = 'throwawaymail';
    private const API_BASE = 'https://throwawaymail.app/api';

    /** @var array<string,string> */
    private const HEADERS = ['Accept' => 'application/json'];

    public static function generate(): EmailInfo
    {
        $resp = HttpClient::post(self::API_BASE . '/mailboxes', self::HEADERS);
        $data = HttpClient::json($resp);
        $mailboxId = trim((string) ($data['mailbox_id'] ?? ''));
        $email = trim((string) ($data['address'] ?? ''));
        if ($mailboxId === '' || $email === '' || !str_contains($email, '@')) {
            throw new \RuntimeException('throwawaymail: 创建邮箱响应无效');
        }
        $exp = $data['expires_at'] ?? null;
        return new EmailInfo(
            self::CHANNEL,
            $email,
            $mailboxId,
            expiresAt: is_numeric($exp) ? (int) $exp : null,
            createdAt: $data['created_at'] ?? null,
        );
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        $mailboxId = trim((string) $token);
        $address = trim($email);
        if ($mailboxId === '') {
            throw new \InvalidArgumentException('throwawaymail: mailbox id 不能为空');
        }
        if ($address === '') {
            throw new \InvalidArgumentException('throwawaymail: 邮箱地址不能为空');
        }

        $resp = HttpClient::get(self::API_BASE . '/mailboxes/' . rawurlencode($mailboxId) . '/messages', self::HEADERS);
        $rows = HttpClient::json($resp);
        if (!array_is_list($rows)) {
            return [];
        }

        $out = [];
        foreach ($rows as $item) {
            if (!is_array($item)) {
                continue;
            }
            $messageId = trim((string) ($item['message_id'] ?? ''));
            $detail = $messageId !== '' ? self::fetchMessage($mailboxId, $messageId) : null;
            $out[] = Normalize::email(self::flatten($detail ?? $item, $address), $address);
        }
        return $out;
    }

    /**
     * @return array<string,mixed>|null
     */
    private static function fetchMessage(string $mailboxId, string $messageId): ?array
    {
        $resp = HttpClient::get(
            self::API_BASE . '/mailboxes/' . rawurlencode($mailboxId) . '/messages/' . rawurlencode($messageId),
            self::HEADERS,
        );
        if ($resp->getStatusCode() >= 400) {
            return null;
        }
        $data = HttpClient::json($resp);
        return !empty($data) ? $data : null;
    }

    /**
     * @param array<mixed> $raw
     * @return array<string,mixed>
     */
    private static function flatten(array $raw, string $recipient): array
    {
        $toAddresses = $raw['to_addresses'] ?? null;
        $to = (is_array($toAddresses) && isset($toAddresses[0])) ? (string) $toAddresses[0] : $recipient;
        return [
            'id' => $raw['message_id'] ?? '',
            'messageId' => $raw['message_id'] ?? '',
            'from_address' => $raw['from_address'] ?? '',
            'fromName' => $raw['from_name'] ?? '',
            'to' => $to,
            'subject' => $raw['subject'] ?? '',
            'received_at' => $raw['received_at'] ?? '',
            'read' => $raw['read'] ?? false,
            'text' => $raw['text'] ?? '',
            'html' => $raw['html'] ?? '',
            'size' => $raw['size'] ?? null,
        ];
    }
}
